### DISCUSSION TITLE
Scripts

### DISCUSSION THREAD

**Discussion by Anprivate** [fedorovvl/tso_client#36] [Q&A]
> support for user scripts (place user.js to tso_portable folder)

О, мечты сбываются, я как раз хотел просить либо какие-нибудь скрипты, либо некий API
А есть какая-нибудь документация?

**Comment by fedorovvl** [comment]
документация на adobe air и jquery есть в общедоступных источниках
as3 код флешки можно почитать открыв флешку в as3sorcerer например
в [user.js](https://github.com/fedorovvl/tso_client/blob/master/client/user.js) я прям кратенько показал как обращаться к методам внутри флешки
через CustomCode при --debug можно кстати тестировать вызов методов или получение данных
важное уточнение. обращаться можно только к public методам, св-вам и переменным

**Reply by Anprivate** [reply]
Ага. То есть вот эти вот запросы вида:
swmmo.application.mGameInterface.mHomePlayer.GetPlayerLevel());
надо вытаскивать непосредственно из флешки. Ок, попробую разобраться.

**Reply by fedorovvl** [reply]
все что после swmmo.application да, это уже классы и методы внутри флешки

**Comment by MkFera** [comment]
I wrote a simple script to get guild members data, so now I can monitoring who is inactive
this help me a lot ! :))))
[user.save.zip](https://github.com/fedorovvl/tso_client/files/9728502/user.save.zip)
internal file must rename to user.js

**Reply by fedorovvl** [reply]
glad to hear! first script from community :) keep going

**Reply by MikaelEngvall** [reply]
Could you add the amount of guild coins received too? I have tried to commit an update with that included but ...

**Reply by MikaelEngvall** [reply]
Would this work?
// ========== GUILD COMMAND

addToolsMenuItem(loca.GetText("LAB", "Guild"), _exudGuildMenuHandler);
var _members = new Array();


function _exudGuildMenuHandler(event) {
	// close all modals
	$("div[role='dialog']:not(#GuildModal):visible").modal("hide");
	// create modal
	createModalWindow('GuildModal', 'Guild Memebrs List');
	if($('#GuildModal .modal-footer ._exudCSVexportBtn').length == 0)
	{
		$('#udGuildStyle').remove();
		if($('#udGuildStyle').length == 0)
		{
			$("head").append($("<style>", { 'id': 'udGuildStyle' }).text('div .row:hover {background-color: #A65329;}'));
		}


		$("#GuildModal .modal-footer").prepend(
			$('<button>').attr({ "id": "_exudCSVexportBtn", "class": "btn btn-primary pull-left _exudCSVexportBtn" }).text("Export CSV")
		);
		$('#GuildModal ._exudCSVexportBtn').click(_exGMCSVexport);
		
		$('#GuildModal .modal-header').append('<div class="container-fluid"><br/>' + createTableRow([
					[4, loca.GetText("LAB", "UserName")],
					[2, loca.GetText("LAB", "GuildQuestStatus")],
					[2, loca.GetText("LAB", "GuildOnlineLast24")],
					[2, loca.GetText("LAB", "GuildLevel")],
					[2, loca.GetText("LAB", "GuildCoins")]
					], true)  + '</div>'
					);			
	}
	
	// fill modal data 
	$('#GuildModalData').html('<div class="container-fluid">{0}</div>'.format(_exudGuildGetData()));
	// show modal
	$("#GuildModal .modal-footer .btn-danger").html(loca.GetText("LAB", "Close"));
	$('#GuildModal:not(:visible)').modal({ backdrop: "static" });

}

function _exGMCSVexport()
{
	var out = 'name;level;quest;active;coins\n';
	_members.forEach(function(m) {
		out += '{0};{1};{2};{3};{4}\n'.format(m.name,m.level,m.quest,m.active,m.coins);
	});
	var file = new air.File(readLastDir('guild')).resolvePath("guild.csv");
	file.addEventListener(air.Event.COMPLETE, function(event){
		saveLastDir('guild', event.target.parent.nativePath);
	}); 
	file.save(out);
}
						
function _exudGuildGetData() {
	
	var out = '';
	var guild = swmmo.application.mGameInterface.GetCurrentPlayerGuild();
	
	if (guild == null || typeof guild == "undefined") { return '<H3>No Data</H3>'; }
	$("#GuildModal .modal-title").text("{0} ( {2}: {1} )".format(guild.name, guild.members.length, loca.GetText("LAB", "GuildMembers")));
	
	_members = new Array();
	for(key in guild.members)
	{
		_members.push({
			"name" : guild.members[key].username,
			"quest" : (guild.members[key].questsStatus == 3 ? loca.GetText("LAB", "Close") : loca.GetText("LAB", "GuildQuestOpen")),
			"active" : (guild.members[key].onlineLast24 ? loca.GetText("LAB", "YES") : loca.GetText("LAB", "NO")),
			"level" : guild.members[key].playerLevel,
			"coins" : guild.members[key].coinsEarned || 0 // Adding coins information, default to 0 if not available
		});
	}
	_members.sort(function(a, b){
		return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
	});
	_members.forEach(function(m) {
		out += createTableRow([
			[4, m.name],
			[2, m.quest],
			[2, m.active],
			[2, m.level],
			[2, m.coins]
		]);
	});
	return out;
}

**Comment by azurpc** [comment]
Cool ! In the user list, could you get member guilde quest status ? thanks

**Reply by BasilMas** [reply]
> Cool ! In the user list, could you get member guilde quest status ? thanks

I've added it by myself modifying table header and adding _questsStatus_ value

The quest status is stored in  _questsStatus_, equil 3 for close and 1 for open guild quest

```
	else
	{
...
		out += '<table><tr><td width="300px">Name</td><td width="100px">GuildQuest</td><td width="100px">Active</td><td width="100px">Level</td></tr>';
...
		{
			var m = guild.members[y];
			out += '<tr>';
			out += '<td>' + m.username + '</td>';
			if (m.questsStatus == 3) {
				out += '<td>' + 'Close' + '</td>';
			}
			else {
				out += '<td>' + 'Open' + '</td>';
			}	
			out += '<td>' + m.onlineLast24 + '</td>';
			out += '<td>' + m.playerLevel + '</td>';
			out += '</tr>';
			y++;
		}
		out += '</table>';
	} 
	
```		

Many thanks @MkFera for example and  @fedorovvl for ditailed explanation

**Reply by azurpc** [reply]
i'm affraid to make errors, could you upload your user.js with your modifications (guilde) ? thanks

**Reply by BasilMas** [reply]
> i'm affraid to make errors, could you upload your user.js with your modifications (guilde) ? thanks

np) File has already proper name 'user.js'
[user.zip](https://github.com/fedorovvl/tso_client/files/9735100/user.zip)

**Reply by skelgaard** [reply]
do it just need to be placed in same folder as the exe file or is there a specific place it needs to be placed ?

**Reply by BasilMas** [reply]
> do it just need to be placed in same folder as the exe file or is there a specific place it needs to be placed ?

it should be placed in tso_portable folder - %localappdata%\temp\tso_portable

**Comment by MkFera** [comment]
I´m writing a command to pickup the collectables , it works fine on home island and adventure, but is not optimized because I don´t hunderstand exactly when is normal building or collectable building, someone can help me?
this is the script :

-------------------------- CUT HERE -----------

addMenuItem("Collectables", CollMenuHandler);

//====== COLLECTABLE COMMAND
function CollMenuHandler(event)
{
	ex_HLColl();
}

function ex_HLColl()
{
	try{
		var _Buildings = swmmo.application.mGameInterface.mCurrentPlayerZone.mStreetDataMap.GetBuildings_vector();
		if (_Buildings == null)
		{
			alert("Nothing to do");
			return;
		}
		var nBuildings = _Buildings.length;
		
		var y = 0;
		while(y < nBuildings)
		{			
			var _loc2_ = _Buildings[y];
			try
			{
                               // check if is collectable building or not
				if (_loc2_.mIsSelectable && !_loc2_.IsInitialSetOnMap() && !_loc2_.IsBuildOnWater())
				{
				_loc2_.handleSelectBuilding(); // is collectable 
				}
			}
			catch (err1)
			{
			}			
			++y;
		}			
	}
	catch (error)
	{
		alert(error.message);
	}
}

**Reply by fedorovvl** [reply]
u need instance of CollectionsManager
```
var collection_manager = swmmo.getDefinitionByName("Collections::CollectionsManager").getInstance();
```
and then check building is collectible
```
collection_manager.getBuildingIsNormalCollectible(_loc2_.GetBuildingName_string())
collection_manager.getBuildingIsEventCollectible(_loc2_.GetBuildingName_string())
```

**Reply by azurpc** [reply]
> [MkFera ](/MkFera) [ il y a 14 heures](#discussioncomment-3834537)

Good job, it's working (collectable quest is ok) but strange thing, no messages when i clic on menu item ? no text, noting happens but collectables ok

**Reply by BasilMas** [reply]
> > [MkFera ](/MkFera) [ il y a 14 heures](#discussioncomment-3834537)
> 
> Good job, it's working (collectable quest is ok) but strange thing, no messages when i clic on menu item ? no text, noting happens but collectables ok

Yes, I confirm it. No messages if I have no collectables (instead "Nothing to do"), and not all collectable were collected if I have them.

**Reply by hoanhtuan-hat** [reply]
I tested, in this case, we need to send a refresh packet (F2) and then run collectible script --> result will appear.

**Reply by Unddann** [reply]
Hello, everyone. Would it be possible to install a timer? So that we can see when the next collectibles will appear?

**Reply by fedorovvl** [reply]
not possible.. client doesn't know when collectibles appears

**Reply by Unddann** [reply]
Could it work with the 8 hour rule? Collectible parts appear every 8 hours.

**Reply by fedorovvl** [reply]
no reference point for timing

**Reply by rrnow** [reply]
no reference point for timing unless you observe one collectibles appearance time and count from it, I think it is the best that can be done.

**Comment by MkFera** [comment]
Thanks fedorovvl  , this works :)

=======
addMenuItem("Collectables", CollMenuHandler);

//====== COLLECTABLE COMMAND
function CollMenuHandler(event)
{
	ex_HLColl();
}


function ex_HLColl()
{
	try{
		var _Buildings = swmmo.application.mGameInterface.mCurrentPlayerZone.mStreetDataMap.GetBuildings_vector();
		if (_Buildings == null)
		{
			alert("Error loading building table !");
			return;
		}
		var nBuildings = _Buildings.length;
		
		var collection_manager = swmmo.getDefinitionByName("Collections::CollectionsManager").getInstance();
		
		var y = 0;
		var collected = 0;
		while(y < nBuildings)
		{			
			var _loc2_ = _Buildings[y];
			try
			{
				if (_loc2_.mIsSelectable && (
						collection_manager.getBuildingIsNormalCollectible(_loc2_.GetBuildingName_string()) ||
						collection_manager.getBuildingIsEventCollectible(_loc2_.GetBuildingName_string())
					)
				)
				{
				_loc2_.handleSelectBuilding();
				++collected;
				}
			}
			catch (err1)
			{
			}
			
			++y;
		}	
		
		alert(collected + " items collected");
		
	}
	catch (error)
	{
		alert(error.message);
	}

}

**Reply by fedorovvl** [reply]
you can also use
```
swmmo.application.mGameInterface.mCurrentPlayerZone.mStreetDataMap.GetBuildings_vector().forEach(function(_loc2_){
});
```
instead of while cycle

**Reply by BasilMas** [reply]
Can we use smth like this for Arctic Explosion to collect boxes/carts?

**Reply by MkFera** [reply]
I´m trying to find them. but they are not in the GetBuildings_vector :(

**Reply by skelgaard** [reply]
is there any documentation for what you can do with the different parts of this ? i can see big potential in this already :)

**Reply by MkFera** [reply]
I also added the search for adventure items like arctic explosion
[user.zip](https://github.com/fedorovvl/tso_client/files/9800050/user.zip)

**Reply by azurpc** [reply]
> I also added the search for adventure items like arctic explosion [user.zip](https://github.com/fedorovvl/tso_client/files/9800050/user.zip)

Thanks a lot !

**Reply by azurpc** [reply]
Hello MkFera, it's possible to have a sorted list of guild members or it's difficult ?

**Reply by MkFera** [reply]
Yes, now it is sorted
[user.zip](https://github.com/fedorovvl/tso_client/files/9810254/user.zip)

**Reply by azurpc** [reply]
1000 thanls Mkfera :)
 A little problem with caps (exemple) : Members Azur, alien, Nobody => sorted : Azur, Nobody, alien instead of alien, Azur, Nobody

**Reply by MkFera** [reply]
can you make the refresh option also reload user.js ? (if it is possible.)
to avoid exiting and reloading everything while developing the script.

**Reply by fedorovvl** [reply]
i think its not possible with javascript
if you load script code you can't unload it

**Reply by MkFera** [reply]
Now sort with ignorecase
[user.zip](https://github.com/fedorovvl/tso_client/files/9812773/user.zip)

**Reply by BasilMas** [reply]
> Thanks fedorovvl , this works :)
> 
> ======= addMenuItem("Collectables", CollMenuHandler);
> 
> //====== COLLECTABLE COMMAND function CollMenuHandler(event) { ex_HLColl(); }
> 
> function ex_HLColl() { try{ var _Buildings = swmmo.application.mGameInterface.mCurrentPlayerZone.mStreetDataMap.GetBuildings_vector(); if (_Buildings == null) { alert("Error loading building table !"); return; } var nBuildings = _Buildings.length;
> 
> ```
> 	var collection_manager = swmmo.getDefinitionByName("Collections::CollectionsManager").getInstance();
> 	
> 	var y = 0;
> 	var collected = 0;
> 	while(y < nBuildings)
> 	{			
> 		var _loc2_ = _Buildings[y];
> 		try
> 		{
> 			if (_loc2_.mIsSelectable && (
> 					collection_manager.getBuildingIsNormalCollectible(_loc2_.GetBuildingName_string()) ||
> 					collection_manager.getBuildingIsEventCollectible(_loc2_.GetBuildingName_string())
> 				)
> 			)
> 			{
> 			_loc2_.handleSelectBuilding();
> 			++collected;
> 			}
> 		}
> 		catch (err1)
> 		{
> 		}
> 		
> 		++y;
> 	}	
> 	
> 	alert(collected + " items collected");
> 	
> }
> catch (error)
> {
> 	alert(error.message);
> }
> ```
> 
> }

It works not only in arctic explosion but in The Lost Skull too for ex

**Reply by azurpc** [reply]
> Now sort with ignorecase [user.zip](https://github.com/fedorovvl/tso_client/files/9812773/user.zip)

well done ;)

**Reply by MkFera** [reply]
> i think its not possible with javascript if you load script code you can't unload it

How load Template and Save Template buttons works ?

I try this but nothing happen

var a = document.createElement("input");
    a.type = "file";
    a.addEventListener('change', LoadSpecDataFile, false);
    a.click();
	delete a;

**Reply by fedorovvl** [reply]
try this
```
$("<button>").text("ButtonText").attr({ "class": "btn btn-success" }).click(LoadSpecDataFile).appendTo("#exampleModal .modal-footer");
```
then in index.html check functions loadTemplate, loadTemplateComplete and loadTemplateLoaded
also readLastDir function if you just want to read file

**Comment by azurpc** [comment]
Hello, I search a user.js script to list all the general (army) of a player (usefull to help player for me to make personnal tutos for players based on their army). Thanks

**Reply by MkFera** [reply]
I´m starting to do someting with specialist classes. 
Take a look with this script
But at the moment I do not know why I´m doing this  ahahahahaha 
this script shows units on the actual zone. so if you are in a firend island, this shows friends specialist. 
in a coop adventure I think it shows all specialist sents for all players.
[user.zip](https://github.com/fedorovvl/tso_client/files/9814258/user.zip)

**Reply by azurpc** [reply]
thanks it's fun

**Reply by azurpc** [reply]
Now, save template of generals ? :)
like explo/geo
you understant where i want to go in final, lol ;)

**Comment by skelgaard** [comment]
@fedorovvl is there any documentation on the different functions that can be used ? am looking into buildings and want different status on them ?

**Reply by fedorovvl** [reply]
nope, only by digging swf =)
open swf with as3sorcerer for example and explore classes
for buildings you need GO.cBuilding class

**Reply by skelgaard** [reply]
ok thanx... i will take a look and see if it can find what i need :)

**Reply by skelgaard** [reply]
trying to do 3 overviews:
- mines overview
- buildings buff overview
- list of buildings that have a buff that can be used, like the christmas marked and so on.....

for now just working with the first one :)

**Reply by azurpc** [reply]
yes looking swf is the only documentation :) you are all wonderful i like the energy of this github. Good Halloween ;)

**Reply by BasilMas** [reply]
> trying to do 3 overviews:
> 
> * mines overview
> * buildings buff overview
> * list of buildings that have a buff that can be used, like the christmas marked and so on.....
> 
> for now just working with the first one :)

may I suggest another task?
To find out the production time of cultural buildings such as Blacktrees, Chocolatiers, Laboratories, etc

They have a queue, and it will be awesome to find out its summary for each building to avoid empty downtime. Unfortunately, my poor knowledge is not enough to do this.
I mean this info - here I have  54 days 15 hours and 52 minutes

![image](https://user-images.githubusercontent.com/76757249/196546290-f8208304-098f-4c71-9434-767974ee3482.png)

**Reply by skelgaard** [reply]
yes may wery well be in this i'm trying to get too.... but first the simple, then the more advanced stuff :)

**Reply by azurpc** [reply]
> yes may wery well be in this i'm trying to get too.... but first the simple, then the more advanced stuff :)

we can have an alpha version of your user.js ?

**Reply by skelgaard** [reply]
> nope, only by digging swf =) open swf with as3sorcerer for example and explore classes for buildings you need GO.cBuilding class

is there any easy way to do a walkthough of them ? to get each of them, cause i got too much on the list i have ;)

**Reply by fedorovvl** [reply]
```
swmmo.application.mGameInterface.mCurrentPlayerZone.mStreetDataMap.GetBuildings_vector().forEach(function(item){
// do work
});
```
item is building (instance of cBuilding)

**Reply by skelgaard** [reply]
well then i have the correct list... was just looking now and saw cDeposit, was thinking that could be used to do the mine list, but unsure...
they seem to be in the other list too

**Reply by skelgaard** [reply]
can't quite figure out how to get what type of building it is (like a mine) and the state of it :)

**Reply by skelgaard** [reply]
@MkFera you got an idea with all your great scripts ;)

**Comment by MkFera** [comment]
A new wip version of my user.js with  "generals" option. load/save send to home , star e adventure


[user.zip](https://github.com/fedorovvl/tso_client/files/9829645/user.zip)

**Reply by azurpc** [reply]
@MkFera : WONDERFUL

**Reply by Unddann** [reply]
Hi. would it be possible to save the required troops for an adventure with the generals? So that, when the file is called up, the generals are loaded correctly and automatically?

**Reply by Unddann** [reply]
Hello MkFera. there are any news of the improvements of the "generals.js"?

**Comment by MkFera** [comment]
Eureka !
I modified the user.js so now I have a "Tools" menu with the scripts. each scripts in a separate file and I can reload them without restart de client.
![image](https://user-images.githubusercontent.com/114618057/197001913-ff190926-b7a3-4e0b-9ee1-66d8cac67adc.png)
[userjs.zip](https://github.com/fedorovvl/tso_client/files/9831983/userjs.zip)

So I can upload only the script with the new function and not all code.

If you not use tools menu, you can copy the content of user_script and put into your  user.js  , looks at the top of the script

Example :
// comment if not use tools menu features
_exudAddSubMenuItem( "Collectables", _exudCollMenuHandler);
// uncomment if not use tools menu
//addMenuItem("Collectables", _exudCollMenuHandler);

**Reply by azurpc** [reply]
@MkFera  : Message from a player, Denis (for info)
the charged generals empty on arrival on the island of adventure, there is also a crash when you enter them all in the star

**Reply by igorSof** [reply]
всё работает, в приключение по шаблону генералы отправляются, вместе с армией. вот бы еще заправлять солдатами по шаблону
everything works, the generals go on an adventure according to the template, along with the army. that would still fill the soldiers according to the template

**Reply by skelgaard** [reply]
i love this addition.... wonder if it some way would be able to auto ddiferent user_ js files or that is @fedorovvl that needs to add that to client, that way different scripts from different user can be loaded without having to edit on central file.

**Reply by fedorovvl** [reply]
im thinking about it.. maybe scripts folder in tso_portable

**Comment by MkFera** [comment]
I removed the setTimeout  and changed the sort , your general first.
Load and save does not record army configuration but only which generals are selected.

[user_generals.zip](https://github.com/fedorovvl/tso_client/files/9833384/user_generals.zip)

**Reply by ElectricImpossible** [reply]
Do you expect to eventually record army configuration?

**Reply by MkFera** [reply]
I already thought about this, but for now I dont think to do it because is a litte complex , many things may changes when reload the template, for example not  sufficient army to fill generals, like saved.

**Reply by hoanhtuan-hat** [reply]
This is really good. I have the idea but don't know how to do it:
- Load army pre-configuration for an adventure, if not enough soldiers then pup up alert, If enough then send to adventure
- Have a menu to do SINGLE step attack:
     - 1. Attack G1. Send Gen X + Gen Y 
     - 2. Attack G3. Send Gen K + Gen Z.....
Player just need to monitor and do step by step

**Comment by MkFera** [comment]
Added sort by name / sort by type in generals.js, and set to Nothing the default send action.

In my user.js script I changed how scripts are loaded. 
Now I put them in Scripts directory and is not more necessary put them into the script user.js. 
Now it  load automatically all .js in Script directory.
So if you add  new script while client is running, the reload option also update the menu with the new script.

*** if you are using classic user.js , remember to uncomment/comment the first 2 lines of my user_xxxx.js scripts when copy the code to your user.js

[userjs.zip](https://github.com/fedorovvl/tso_client/files/9843156/userjs.zip)

**Reply by fedorovvl** [reply]
join to our discord server https://discord.gg/YTjs3cUk
i need to show you new code version

**Reply by skelgaard** [reply]
to bad its not on english.... most of us will not understand what you write in there, when not using english

**Reply by fedorovvl** [reply]
there is a room for non russian speaking, so we can easily speak on english there

**Reply by skelgaard** [reply]
i'm in there and only se one none english room on it

**Reply by skelgaard** [reply]
ah needed to press a button which i couldn't read what was on it, to get more rooms... gave more rooms i don't understand :)

**Reply by azurpc** [reply]
> Added sort by name / sort by type in generals.js, and set to Nothing the default send action.
> 
> In my user.js script I changed how scripts are loaded. Now I put them in Scripts directory and is not more necessary put them into the script user.js. Now it load automatically all .js in Script directory. So if you add new script while client is running, the reload option also update the menu with the new script.
> 
> *** if you are using classic user.js , remember to uncomment/comment the first 2 lines of my user_xxxx.js scripts when copy the code to your user.js
> 
> [userjs.zip](https://github.com/fedorovvl/tso_client/files/9843156/userjs.zip)

Great job as usually :)
I've ideas : a button 'show only selected generals", usefull to verify general before sending
'hide guest general', it's a bit anoying to see them on venture
I 've a strange bug (see picture)
![bug](https://user-images.githubusercontent.com/109301910/197350326-bee95675-dd79-499e-9135-466859725be8.jpg)
Thanks !

**Reply by Oviz** [reply]
i want this script too :)

**Reply by skelgaard** [reply]
just download the userjs.zip, unpack them into a userscripts folder next to your client.exe file and they will work :)
ah can se its an outdated zip file... will get a new one and upload

**Reply by skelgaard** [reply]
[userscripts.zip](https://github.com/fedorovvl/tso_client/files/9851261/userscripts.zip)

**Comment by Joahim49** [comment]
Hi, all! How to get a server name in tittle?

**Reply by fedorovvl** [reply]
```
var global = swmmo.getDefinitionByName("global");
document.title += ' - ' + global.gameworld;
```

**Comment by Joahim49** [comment]
Thx 😉 ;*


Wysłane z aplikacji Poczta dla systemu Windows 

Od: fedorovvl
Wysłano: niedziela, 23 października 2022 15:51
Do: fedorovvl/tso_client
DW: Joahim49; Comment
Temat: Re: [fedorovvl/tso_client] Scripts (Discussion #36)

var global = swmmo.getDefinitionByName("global");
document.title += ' - ' + global.gameworld;

**Comment by AceOfSettlers** [comment]
Hello everybody, it's perfect that now we can see before we send GEO and EXPLO at what time (date) they will come back , but is it possible to have a script (user file) that we can put on script directory that will show a window were we can check in a list at what time Explo ... GEO will come back , because now we have this information only at the time of sendind the explo .. Geo , I don"t know if my request is clear ? :)

**Comment by RedzIta** [comment]
Hi all, I've created a new script to display all the buildings on the island, but at the moment I can't:
- view the level reached by the building
- filter the different types
- translate all buildings into the player's language
- understand if a building is on or off

Can you help me?
Attached you will find the first version.

[user_buildings.zip](https://github.com/fedorovvl/tso_client/files/10254660/user_buildings.zip)

**Reply by mcr120** [reply]
Are you planning to implement a mass buff function for buildings ?

**Reply by fedorovvl** [reply]
for level use item.GetUpgradeLevel()
for on/off use item.IsProductionActive()
filter by item.IsDecoration() and check item.GetGOContainer()....

**Reply by skelgaard** [reply]
@RedzIta did you get this to work ? is there a newer version of this ?

**Reply by RedzIta** [reply]
This is the new version
[user_buildings.zip](https://github.com/fedorovvl/tso_client/files/10437198/user_buildings.zip)

@fedorovvl you can add the script in the next release.

**Reply by skelgaard** [reply]
@RedzIta i like your script, but it shouldn't add an item in dropdown for each building type. thats what you get when you choose the building type  :)

**Reply by RedzIta** [reply]
@skelgaard, my idea was grouping all buildings by type and level and for single result the total of buildings, so you can watch quickly if you must upgrade a building type.

**Reply by skelgaard** [reply]
@RedzIta hmmm good point... 
but there is also a problem with that, like blacktree or lovetree... i would newer upgrade all of them, so that makes it have more entries of them alwas in the list.
It would also be a good idea, if it was showing if the building was upgraded, in the specific list.

**Comment by RedzIta** [comment]
As you can see in the image the info present (highlighted in the red rectangle)

![image](https://user-images.githubusercontent.com/120989098/219977463-80f0549d-1c16-40e2-80a9-e0cd5f63e6b8.png)

**Comment by Glebsky** [comment]
[index.html](https://github.com/user-attachments/files/21820785/index.html)
Game Assest and their ids

**Comment by Glebsky** [comment]
Link for the web version of icon in game and theirs ids.
[Icons](https://htmlpreview.github.io/?https://github.com/Glebsky/tso_client/blob/icons-feature/library/tets.html)

**Comment by Glebsky** [comment]
[lang.txt](https://github.com/user-attachments/files/23147543/lang.txt)
ru lang file

**Reply by igorSof** [reply]
А куда положить этот файл? что бы по-русски стало всё, а то много надписей типа такой "NT:Hepabedunavo"

**Reply by igorSof** [reply]
вот бы кто еще перевел его, а то там за 3 года без переводов.

**Reply by Glebsky** [reply]
This is for creating scripts — it’s meant to make it easier to find the necessary phrases. You won’t be able to translate the game this way, since the language files are loaded directly from the game server.

**Reply by Glebsky** [reply]
> вот бы кто еще перевел его, а то там за 3 года без переводов.

там кстати вышел скрипт, который обновляет, недостающие переводы.

**Comment by Glebsky** [comment]
[client_scripts.txt](https://github.com/user-attachments/files/25204208/client_scripts.txt)
client source code

**Reply by Oviz** [reply]
how build?

**Reply by Glebsky** [reply]
It is just code, To find functionality for your scripts

**Comment by Glebsky** [comment]
[icons_resource.html](https://github.com/user-attachments/files/25385771/icons_resource.html)
Resource icons/images.

**Comment by krisrmgua** [comment]
I have a script that will open the bookbinder is there a way to have it collect the manuscript button and the click the manuscript and accept to create a new one?

```// Bookbinder - Open the Bookbinder building

addToolsMenuItem(loca.GetText("BUI", "Bookbinder"), bookbinderOpenHandler);

function bookbinderOpenHandler(event) {
    try {
        if (!game.gi || !game.gi.isOnHomzone || !game.gi.isOnHomzone()) {
            game.showAlert(getText('not_home'));
            return;
        }
        var vector = game.gi.mCurrentPlayerZone.mStreetDataMap.getBuildingsByName_vector("Bookbinder");
        if (!vector || vector.length === 0) {
            game.showAlert("Bookbinder not found.");
            return;
        }
        vector.sort(function(a, b) { return a.GetUpgradeLevel() > b.GetUpgradeLevel() ? -1 : 1; });
        game.zone.ScrollToGrid(vector[0].GetGrid());
        game.gi.SelectBuilding(vector[0]);
    } catch (e) {
        game.showAlert("Bookbinder error: " + (e.message || e));
    }
}
```

**Comment by jackrtenboer-sys** [comment]
I have a question for a userscript: Can a script be made for reading the mails from explorers and export the content to csv for examining?
Thanx in advance for an answer

**Comment by Glebsky** [comment]
🎨 TSO Icon Finder — быстрый поиск и скачивание почти 8 000 иконок из The Settlers Online. Поддержка XML-тегов, удобный зум и превью. Незаменимо для гайдов, скриптов, сайтов и фанатских проектов!

🎨 TSO Icon Finder — browse, preview, and download nearly 8,000 The Settlers Online icons with instant search and XML tag support. Perfect for guide, scripts, creators and fan sites!

👉 https://tso-icon-finder.vercel.app/
